File: algorithms/trees/fenwick-tree/csharp/FenwickTree.cs
using System;

class FenwickTree
{
    private int[] tree;
    private int n;

    public FenwickTree(int[] arr)
    {
        n = arr.Length;
        tree = new int[n + 1];
        for (int i = 0; i < n; i++)
            Update(i, arr[i]);
    }

    public void Update(int i, int delta)
    {
        for (++i; i <= n; i += i & (-i))
            tree[i] += delta;
    }

    public int Query(int i)
    {
        int sum = 0;
        for (++i; i > 0; i -= i & (-i))
            sum += tree[i];
        return sum;
    }

    static void Main(string[] args)
    {
        int[] arr = { 1, 2, 3, 4, 5 };
        var ft = new FenwickTree(arr);
        Console.WriteLine("Sum of first 4 elements: " + ft.Query(3));

        ft.Update(2, 5);
        Console.WriteLine("After update, sum of first 4 elements: " + ft.Query(3));
    }
}

File: algorithms/trees/fenwick-tree/python/FenwickTree.py
class FenwickTree:
    def __init__(self, arr):
        self.n = len(arr)
        self.tree = [0] * (self.n + 1)
        for i, v in enumerate(arr):
            self.update(i, v)

    def update(self, i, delta):
        i += 1
        while i <= self.n:
            self.tree[i] += delta
            i += i & (-i)

    def query(self, i):
        s = 0
        i += 1
        while i > 0:
            s += self.tree[i]
            i -= i & (-i)
        return s


if __name__ == "__main__":
    arr = [1, 2, 3, 4, 5]
    ft = FenwickTree(arr)
    print(f"Sum of first 4 elements: {ft.query(3)}")

    ft.update(2, 5)
    print(f"After update, sum of first 4 elements: {ft.query(3)}")
